### PR TITLE
0.8 Pre-release (Docs, tutorial, etc)

### PR DESCRIPTION
**Pull request by jojolepro**
Rebasing after the review phase.

System and Dispatcher pages made by moxinillian (THANKS!)

Enjoy the 5000 words tutorial ;)

---
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/832)

### CODE REVIEW

File: book/src/getting-started.md [hunk 1]
```diff
@@ -13,25 +13,31 @@ If you are on Linux, you'll need to install `libasound2-dev` and `libx11-xcb-dev
...
+### Cargo (Manual)
+
 In case you're doing this with `cargo`, here's what you need to do:
 
 * Add `amethyst` as dependency in your `Cargo.toml`.
 * Create a `resources` folder and put a `display_config.ron` in it.
-* Start with one of the [examples][ex] from the Amethyst repository (e.g. `window`)
-  for the source code. Watch out to use the right example for the version of Amethyst
-  you specified in `Cargo.toml`.
+* (Optionnal) Copy the code from one of amethyst's examples.
```
Comment (ishitatsuyuki) [inline book/src/getting-started.md]: Typo

**Comment by Moxinilian** [comment]
You see, tutorials are common. They try to teach you stuff, make you learn something. I have seen many tutorials, many great tutorials even. But this one goes far beyond anything I have ever witnessed.

Everything in this piece is fantastic, from the bold and engaged layout, to the subtle and always beautiful examples of Jojo le professionnel's genius mind. I was blown away by the quality of the language, most notably in the System and Dispatcher pages which I think are unforgettable pieces of modern literature: the rest of the book is absolutely fantastic, and yet it sounds poor next to this fabulous work of unprecedented art. From the beginning to the end, I kept being amazed by how brilliantly Joël transmitted the power of his mind and with such ease he returns the way one think about themselves.

Because this is what this amethyst tutorial really is about: getting your understanding of yourself, the human mind, and even the entire universe struck by a few of his words. This touching *œuvre d'art* makes you reconsider everything you ever thought you knew. 

It is more than just a tutorial. During my entire reading, on every single instant, I felt it: the strength of the pong tutorial with the elegance of the `ProgressCounter`'s API reference documentation. After reading this, the old Moxinilian was gone. This piece changed me, I am now a new person.

**TL;DR** yeah good stuff m8

**Comment by minecrawler** [comment]
**I love it.** Especially the system chapter explains easy solutions to real problems when dealing with entities (my pain points from some time ago). Things I would love to see in the docs: saving and loading a game world, with gltf data, ecs components, resources and the kitchen sink. I read a lot of issues, and they are all closed, but I never found a piece of code demonstrating how to actually do so...

All in all, I get the feeling that the engine's basic features are complete, and only the fancy stuff is missing. Way to go!

File: book/src/concepts/system.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+However, this approach is pretty rare because most of the time you don't know what entity you want to manipulate, and in fact you may want to apply your changes to multiple entities.
+
+### Getting all entities with specific components
+
+Most of the time, you will want to perform logic on all entities with a specific components, or even all entities with a selection of components.
+
+This is possible using the `join` method. You may be familiar with joining operations if you have ever worked with databases. The `join` method takes multiple storages, and iterates over all entities that have a component in each of those storages.
+It works like an "AND" gate. It will return an iterator containing a tuple of all the requested components if they are **ALL** on the same entity.
+
+If you join with components A,B and C, only the entities that have **ALL** those components will be considered.
```
Comment (Moxinilian) [inline book/src/concepts/system.md]: Missing space between `A,` and `B`.

File: book/src/concepts/system.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        }
+    }
+}
+```
+
+This system does the same thing as the previous `MakeObjectsFall`, but also cleans up falling objects that reached the ground.
+
+### Adding or removing components
+
+You can also insert or removes components from a specific entity.
```
Comment (Moxinilian) [inline book/src/concepts/system.md]: Too many `s` in `removes`.

File: book/src/concepts/system.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        if data.baz.should_process() {
+            for (foo, mut bar) in (&data.foo, &mut data.bar) {
+                bar.stuff += foo.stuff;
+            } 
+        }
+    }
+}
+```
+
+## The setup method
```
Comment (Moxinilian) [inline book/src/concepts/system.md]: Wouldn't we want the setup method to be before the SystemData trait?
It feels more useful than that trait to me.

File: book/src/concepts/system.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+Systems have a method called setup which is called a single time, before the any of the system runs.
```
Comment (Moxinilian) [inline book/src/concepts/system.md]: before ~~the~~ any of the system runs

File: book/src/getting-started.md (hunk 1 above)
Comment (jojolepro) [reply book/src/getting-started.md]: sorry, but where?

Comment (omni-viral) [reply book/src/getting-started.md]: Optio**nn**al

Comment (jojolepro) [reply book/src/getting-started.md]: Thanks! I thought it was written like that :joy:

File: book/src/concepts/dispatcher.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+When a dispatcher is created, it is associated with the systems that it will execute. It then generates an execution plan that respects mutability rules while maximizing parallelism.
+
+## Respecting mutability rules
+
+When a system wants to access a `Storage` of a resource, they can do so either mutably or immutably. This works just like in Rust: either only one system can request something mutably and no other system can access it, or multiple systems can request something but only immutably.
+
+The dispatcher looks at all the `SystemData` in the systems and builds execution stages.
+
+If you want to have the best performance possible, you should prefer immutable over mutable whenever it is possible. (`Read` instead of `Write`, `ReadStorage` instead of `WriteStorage`)
```
Comment (Xaeroxe) [inline book/src/concepts/dispatcher.md]: This may encourage users to use types with interior mutability "for performance".  That would defeat the purpose of the `ECS`.  One of the primary points is that it doesn't use locking.

Maybe add a note like "We also discourage the use of `Mutex` or `RwLock` in your components or resources, `Write` is preferable to interior mutability."

File: book/src/concepts/dispatcher.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+## Execution stages
```
Comment (Xaeroxe) [inline book/src/concepts/dispatcher.md]: We should probably cut this section, new users don't need to know quite this much about the `Dispatcher` to get started with Amethyst.

File: book/src/concepts/entity.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
+# Entity
+
+## What is an entity?
+
+An `Entity` is simply a collection of `Component`. At least conceptually. It represent a single object in your world.
```
Comment (Xaeroxe) [inline book/src/concepts/entity.md]: s/\`Component\`/\`Component\`s/g

s/represent/represents/g

File: book/src/concepts/entity.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+## Creating an entity
+
+There are two common ways to create entities:
+* From a `World` instance. See the relevant chapter in the book.
+* From a `System` using `Entities`. See the system chapter in the book.
+
+## Getting components of an entity
+
+You can't! Well, at least not directly from an `Entity` instance.
+As mentionned in the component book page, `Component`s are not directly attached to entities; they are inserted into storages.
```
Comment (Xaeroxe) [inline book/src/concepts/entity.md]: s/mentionned/mentioned/g

File: book/src/concepts/event-channel.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
+# Event Channel
+
+This chapter will be easier than the previous ones. 
+
+While it is not essential to understand it to use amethyst, it can make your life much much easier in a lot of situations where using only data would make your code too complex.
+
+## What is an event channel?
+
+An `EventChannel` acts like a queue for any type that implements `Event`.
```
Comment (Xaeroxe) [inline book/src/concepts/event-channel.md]: s/\`Event\`/\`Send + Sync + 'static\`/g

File: book/src/concepts/event-channel.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+The `Event` type is automatically implemented for all types that are `Send`, `Sync` and have a static lifetime.
```
Comment (Xaeroxe) [inline book/src/concepts/event-channel.md]: Omit this line

File: book/src/concepts/event-channel.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+It is a single producer/multiple receiver. This means that it should be used when you only have a single "thing" (usually a system) producing events.
```
Comment (Xaeroxe) [inline book/src/concepts/event-channel.md]: Not sure I agree with this advice.  Multiple producers of events is fine, but it should be understood that they can't all write events in simultaneously.  Multiple reading is definitely simultaneous though.

File: book/src/concepts/event-channel.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+In most case, the `EventChannel` should be stored in a global resource for ease of access. More on this later.
```
Comment (Xaeroxe) [inline book/src/concepts/event-channel.md]: s/case/cases/g

File: book/src/concepts/event-channel.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+
+The only exception to this is when you have a **SINGLE** external thread producing the events and you want your game to have a read access over the `EventChannel`.
```
Comment (Xaeroxe) [inline book/src/concepts/event-channel.md]: I'd omit this line and the next.  This advice focuses on such a narrow use case it might confuse readers.

File: book/src/concepts/event-channel.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+In this specific case, you can create a resource holding a Arc<Mutex<EventChannel<T>>>.
+
+If your use case involves multiple event producers, you should use some other type of event queuing method.
```
Comment (Xaeroxe) [inline book/src/concepts/event-channel.md]: I'd omit this paragraph.  `shrev` works great for multiple event producers, it just comes with some caveats that I've stated above.

File: book/src/concepts/event-channel.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+```rust,ignore
+    let mut reader = channel.register_reader();
+```
+
+Then, when you want to read the events:
+
+```rust,ignore
+    for event in channel.read_event(&mut reader){
+        // The type of the event is infered from the generic type
```
Comment (Xaeroxe) [inline book/src/concepts/event-channel.md]: s/infered/inferred/g

File: book/src/concepts/event-channel.md [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+In those case, you can use a `System` reading from a [mpsc channel](https://doc.rust-lang.org/std/sync/mpsc/) and moving the events into a `EventChannel`.
+You can also use other rust crates for that purpose.
+
+## Creating an event channel
+
+Super simple!
+
+```rust,ignore
+    // In the following examples, we are going to use i32 values as events.
+    let mut channel = EventChannel::<i32>::new();
```
Comment (Xaeroxe) [inline book/src/concepts/event-channel.md]: 9 times out of 10 events should be some kind of enum.  I know this is just example code, but it might accidentally encourage bad practice.  Let's demonstrate this with a simple enum instead.

File: book/src/concepts/event-channel.md [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+It is the `ReaderId` that needs to be mutable to keep track of where your last read was.
+
+## Patterns
+
+When using the event channel, we usually re-use the same pattern over and over again to maximize parallelism.
+It goes as follow:
+
+Create a resource holding the event channel:
+```rust,ignore
+struct MyEventChannel {
```
Comment (Xaeroxe) [inline book/src/concepts/event-channel.md]: I'm not sure this is great practice, adding the raw `EventChannel<Event>` can be a great idea if the Event type is unique enough.  We should encourage using a unique and stronger Event type instead.

File: book/src/concepts/resource.md [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+```rust,ignore
+let mut fetched = resources.try_fetch_mut::<MyResource>().expect("No MyResource present in Resources");
+```
+
+Other ways of fetching a resource will be covered in the system section of the book.
+
+## Deleting a resource
+
+At the time of writing this book, you can't.
+If you have a use case where you need to delete a resource, you probably need to use a custom `GameData` or multiple `World`.
```
Comment (Xaeroxe) [inline book/src/concepts/resource.md]: This is pretty bad advice in my opinion.  Just encourage users to wrap their resource in an `Option` instead and set it to `None` when they want to delete it.

File: book/src/concepts/state.md [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+In the case of amethyst, it is used to represent the "game stage".
+
+A game stage is a *general* and *global* section of the game.
+
+## Example
+
+Let's go with an example.
+You are making a pong game.
+
+* When the user opens up the game, it first loads all the assets and shows a pretty loading screen.
```
Comment (Xaeroxe) [inline book/src/concepts/state.md]: omit pretty here, it's just black and says "Loading".  Not quite inspiring :P

File: book/src/concepts/system.md [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+
+// Remove the component
+write_storage.remove(entity);
+```
+
+Keep in mind that inserting a component on an entity that already has a component of the same type **will overwrite the previous one**.
+
+## The SystemData trait
+
+While this is a rarely useful, it is possible to create custom `SystemData` types.
```
Comment (Xaeroxe) [inline book/src/concepts/system.md]: s/this is a/this is/g

File: book/src/concepts/world.md [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+       .build();
+```
+
+Internally, the `World` interacts with `EntitiesRes`, which is a resource holding the entities inside of `Resources`.
+
+## Accessing a `Component`
+
+```rust,ignore
+    // Create an `Entity` with `MyComponent`.
+    // `World` will implicitely write to the component's storage in `Resources`.
```
Comment (Xaeroxe) [inline book/src/concepts/world.md]: s/implicitely/implicitly/g

File: book/src/getting-started.md [hunk 23]
```diff
@@ -13,25 +13,31 @@ If you are on Linux, you'll need to install `libasound2-dev` and `libx11-xcb-dev
...
 
-You can either use the [Amethyst CLI][cl] or just cargo to set up your project.
-After executing
+You can either use the [Amethyst CLI][cl] or cargo to set up your project.
 
+### Amethyst CLI (Easiest)
+If you wish to use the Amethyst cli tool, you can install it like so
+
+```norun
+cargo install amethyst_cli
```
Comment (Xaeroxe) [inline book/src/getting-started.md]: s/amethyst_cli/amethyst_tools/g

File: book/src/concepts/dispatcher.md (hunk 6 above)
Comment (Moxinilian) [reply book/src/concepts/dispatcher.md]: I think that then it should explain really quickly why it's still better.
I would have went for something more like "Please however keep in mind that `Write` is still preferable to interior mutability, such as `Mutex` or `RwLock` for example."

File: README.md [hunk 24]
```diff
@@ -17,20 +17,44 @@
...
+
+By design, the amethyst engine encourages you to write clean and reusable code for your behaviours and data structures, allowing engine users to easily
+share useful components, thus reducing development time and cost.
+
+Using the [ecs] architecture, the code of games can be cleanly divided between data and behaviour, making it easy to understand what is going on,
+even if the game is running on a massive 64 threads processor.
+
+### Community
+
+While we may not be feature-packed (yet!), we all strongly believe that the community-oriented side of amethyst will make it thrive forward!
```
[the remark this answers is not in the document]
Comment (jojolepro) [reply README.md]: I said
community->engine
you said
engine->community

I did change it to your proposal though.

File: book/src/concepts/component.md [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+
+A component is any struct that can be "attached" to an `Entity` (we will cover entity in the next chapter).
+
+## Usage
+
+The relationship between an entity and a component closely ressembles the relation between a real-life object and its properties.
+
+For example, a bottle of water has a shape, a volume, a color and is made of a material(usually plastic).
+
+In this example, the bottle is the entity, and the properties are components.
```
[the remark this answers is not in the document]
Comment (jojolepro) [reply book/src/concepts/component.md]: I was missing so much inspiration when writing this lol :joy:

File: book/src/concepts/component.md [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+
+`Component`s, in contrast with popular belief, should not be stored directly inside of a `Entity`.
+
+They are instead stored in different types of `Storage`, which all have different performance strategy.
+
+When implementing `Component` for a type, you have to specify which storage strategy it should use.
+
+Here's a comparison of the most used ones:
+* `VecStorage`: Elements are stored into a contiguous array. Uses more memory than `DenseVecStorage` but is more performant when you use a `Component` type a lot.
+* `DenseVecStorage`: Elements are stored in a contiguous array. No empty space is left between `Component`s, allowing a lowered memory usage.
```
[the remark this answers is not in the document]
Comment (jojolepro) [reply book/src/concepts/component.md]: Changed, when I push can you verify that my change corresponds to what you expect?

[the remark this answers is not in the document]
Comment (Xaeroxe) [reply book/src/concepts/component.md]: Of course!

**Comment by jojolepro** [comment]
Looks all good on my side. Clear for everyone?

**Comment by jojolepro** [comment]
bors r+

This helps ;)
